plugins/elasticsearch/elasticsearch.go

```diff
@@ -166,43 +212,66 @@ func (o *ElasticSearchOutput) receiver(or OutputRunner, wg *sync.WaitGroup) {
 				outBatch = append(outBatch, outBytes...)
 				if count = count + 1; o.bulkIndexer.CheckFlush(count, len(outBatch)) {
 					if len(outBatch) > 0 {
-						// This will block until the other side is ready to accept
-						// this batch, so we can't get too far ahead.
 						o.batchChan <- outBatch
 						outBatch = <-o.backChan
-						count = 0
+						atomic.AddInt64(&o.processMessageCount, int64(count))
 					}
+					count = 0
 				}
 			}
 		case <-ticker:
 			if len(outBatch) > 0 {
+				if o.conf.UseBuffering {
+					if err = o.bufferedOut.RollQueue(); err != nil {
+						or.LogError(err)
+						return
```

Comment:
I know you copied this from the TcpOutput, but I'm realizing that it's a bit off. First, an error returned from Run will automatically get logged, so there's no need to log it inside the Run method. Second, that error will be a low level one returned from an os.OpenFile call, so we should probably decorate it a bit to add some context for the poor folks who have to figure out why Heka just shut down on them. I recommend replacing both of these lines with:

 `return fmt.Errorf("can't create new buffer queue file: %s", err.Error())`